File: OpenEdXMobile/src/main/java/org/edx/mobile/module/registration/model/RegistrationFormField.java
package org.edx.mobile.module.registration.model;

import androidx.annotation.NonNull;

import com.google.gson.annotations.SerializedName;

import java.util.ArrayList;
import java.util.List;

public class RegistrationFormField {
    private @SerializedName("required")         boolean required;
    private @SerializedName("name")             String name;
    private @SerializedName("placeholder")      String placeholder;
    private @SerializedName("defaultValue")     String defaultValue;
    private @SerializedName("restrictions")     RegistrationRestriction restriction;
    private @SerializedName("errorMessages")    ErrorMessage errorMessage;
    private @SerializedName("instructions")     String instructions;
    private @SerializedName("type")             RegistrationFieldType fieldType;
    private @SerializedName("label")            String label;
    private @SerializedName("options")          List<RegistrationOption> options; // may be empty
    private @SerializedName("defaultOption")    RegistrationOption defaultOption; // may be nil
    private @SerializedName("supplementalText") String supplementalText;
    private @SerializedName("supplementalLink") String supplementalLink;

    public boolean isRequired() {
        return required;
    }

    public String getName() {
        return name;
    }

    public String getPlaceholder() {
        return placeholder;
    }

    public RegistrationOption getDefaultOption() {
        return defaultOption;
    }

    public RegistrationRestriction getRestriction() {
        return restriction;
    }

    public ErrorMessage getErrorMessage() {
        return errorMessage;
    }

    public String getInstructions() {
        return instructions;
    }

    @NonNull
    public RegistrationFieldType getFieldType() {
        return fieldType != null ? fieldType : RegistrationFieldType.UNKNOWN;
    }

    public String getLabel() {
        return label;
    }

    public List<RegistrationOption> getOptions() {
        if (options == null) {
            options = new ArrayList<>();
        }
        return options;
    }

    public String getDefaultValue() {
        return defaultValue;
    }

    public String getSupplementalText() {
        return supplementalText;
    }

    public String getSupplementalLink() {
        return supplementalLink;
    }
}

File: OpenEdXMobile/src/main/java/org/edx/mobile/module/registration/model/RegistrationFormField.kt
package org.edx.mobile.module.registration.model

import com.google.gson.annotations.SerializedName

data class RegistrationFormField(
    @SerializedName("required")
    val isRequired: Boolean = false,

    @SerializedName("name")
    val name: String? = null,

    @SerializedName("placeholder")
    val placeholder: String? = null,

    @SerializedName("defaultValue")
    val defaultValue: String? = null,

    @SerializedName("restrictions")
    val restriction: RegistrationRestriction = RegistrationRestriction(),

    @SerializedName("errorMessages")
    val errorMessage: ErrorMessage? = null,

    @SerializedName("instructions")
    val instructions: String? = null,

    @SerializedName("type")
    val fieldType: RegistrationFieldType? = RegistrationFieldType.UNKNOWN,

    @SerializedName("label")
    val label: String? = null,

    @SerializedName("options")
    var options: List<RegistrationOption> = arrayListOf(),

    @SerializedName("defaultOption")
    val defaultOption: RegistrationOption? = null,

    @SerializedName("supplementalText")
    val supplementalText: String? = null,

    @SerializedName("supplementalLink")
    val supplementalLink: String? = null,

    @SerializedName("exposed")
    val isExposed: Boolean = false,
) {
    val isEmailField
        get() = RegistrationFieldType.EMAIL.name.equals(name, ignoreCase = true)
    val isConfirmEmailField
        get() = RegistrationFieldType.CONFIRM_EMAIL.name.equals(name, ignoreCase = true)
    val isPasswordField
        get() = RegistrationFieldType.PASSWORD.name.equals(name, ignoreCase = true)
    val isMultiField
        get() = RegistrationFieldType.MULTI.name.equals(name, ignoreCase = true)
}
